orttraining/orttraining/training_api/module.cc

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/execution_provider.h"
+#include "core/session/inference_session.h"
+#include "core/session/environment.h"
+
+#include "orttraining/training_api/include/module.h"
+#include "orttraining/training_api/include/utils.h"
+
+using namespace onnxruntime;
+
+namespace onnxruntime {
+namespace training {
+namespace api {
+
+namespace {
+
+// TODO: consolidate with frontend tooling
+const std::string ACCUMULATE_GRAD_CONTROL_INPUT_NAME{"lazy_reset_grad"};
+
+}  // namespace
+
+Status Parameter::SetGrad(const std::string& gradient_name, const OrtValue& param_grad) {
+  // assert param is allocated
+  ORT_ENFORCE(data_.IsAllocated(), "Parameter data should be allocated before allocating gradient.");
+  ORT_ENFORCE(requires_grad_, "Gradient should only be allocated for trainable parameters.");
+
+  gradient_name_ = gradient_name;
+  gradient_ = param_grad;
+  return Status::OK();
+}
+
+Status Parameter::ResetGrad() {
+  if (!requires_grad_) {
+    return Status::OK();
+  }
+  Tensor* p_tensor = gradient_.GetMutable<Tensor>();
+  const auto& device = p_tensor->Location().device;
+  if (device.Type() == OrtDevice::CPU) {
+    memset(p_tensor->MutableDataRaw(), 0, p_tensor->SizeInBytes());
+  }
+#if defined(USE_CUDA) || defined(USE_ROCM)
+  else if (device.Type() == OrtDevice::GPU) {
+    ORT_NOT_IMPLEMENTED("Not implemented.");
```

Comment:
Does not look right. Do we even need this now given we have lazy reset grad?

Reply:
I think this was an API that we had exposed for `Parameter` but I think it is no longer relevant given `Module::ResetGrad` does not call `Parameter::ResetGrad`. We can remove this method. Will make the change.

Reply:
The other thing to note is that the Module::ResetGrad does only lazy reset, so we might want to expose a way that does explicit reset. This might be a good place to do that. I think we should keep the ResetGrad here and document what it does (and also add support for GPU).